### ISSUE TITLE
Feature request: Concurrently serving models with optimizers

### ISSUE THREAD

**Issue by rothn** [tensorflow/tensorflow#19825]
==Problem==

I want to use TensorFlow Serving to serve an output given certain placeholder inputs with a single run. The graph looks like this:

Input vector -> embedding lookup -> iterative inference with optimizer and tf.while_loop -> matrix math using inferred vector -> top K -> output

The problem is that I can't simply use a Tensor in the loop because tf.Optimizer requires a Variable. This in turn forces me to modify global state in each prediction, effectively prohibiting concurrency.

==Feature Request==

This could be solved by implementing:
* a version of Optimizer.minimize() that returns a list of Tensors and takes Tensors instead of Variables as input and
* an optimization to make Optimizer.minimize() recycle Tensors that are loop variables re-assigned to the return of Optimizer.minimize()

==For tensorflowbutler==

Since this is a feature request, the following fields are irrelevant:
Have I written custom code: NA
OS Platform and Distribution: NA
TensorFlow installed from: NA
TensorFlow version: NA
Bazel version: NA
CUDA/cuDNN version: NA
GPU model and memory: NA
Exact command to reproduce: NA

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by rothn** [comment]
Since this is a feature request, the following fields are irrelevant:
Have I written custom code: NA
OS Platform and Distribution: NA
TensorFlow installed from: NA
TensorFlow version: NA
Bazel version: NA
CUDA/cuDNN version: NA
GPU model and memory: NA
Exact command to reproduce: NA

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 150 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 165 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 180 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gadagashwini** [comment]
Hi @rothn, `tf.Optimizer.minimize()` ,this API returns an Operation that updates the variables in var_list. The iterations will be automatically increased by 1. var_list can converted to Tensors. Recommend you to try the latest Tf-serving and Tensorflow version. Thank you!

**Comment by rothn** [comment]
Thanks. I tend to use TFX lately with an AutoML product that generates the pipeline, so I do not anticipate having the opportunity to try this out soon. Thanks for your response and I will let you know if this works for me when I next need it :-).

**Comment by sushreebarsa** [comment]
Hello @rothn! Could you try to make a custom model and try  with the latest  TF serving. Please refer to[ this ](https://www.tensorflow.org/tfx/serving/custom_op)link for more information on Serving custom ops and let us know if that helps?
Thank you!

**Comment by sachinprasadhs** [comment]
@rothn , While I'm trying to understand where this feature request would fit better so that right team can look into this in detail, do you think this could be part of [tf.keras.Optimizer](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers)?

**Comment by rothn** [comment]
Yes! I think this would be right for tf.keras.Optimizer! Thanks for addressing this.

While I no longer need this, I believe the use-case is valuable. In the description above, I was trying to combine embedding-based information retrieval (which quickly degrades in quality after the first few results) with an online collaborative filtering algorithm trained on a users's history and heavily biased towards top retrieval results.

This involves performing a gradient descent as part of inference, and I suspect this is only part of a larger class of solutions that would benefit from this capability.

Thanks!

**Comment by hertschuh** [comment]
@rothn ,

For this use case, it is actually much easier to not use the `tf.keras.Optimizer` machinery and instead simply have a stateless function that applies the gradient in the way the optimizer would. The `tf.keras.Optimizer`s maintain different kinds of states for different reasons, which doesn't work for your use-case.

```py
def apply_sgd(gradient, tensor):
  tensor = tensor - gradient * learning_rate
  return tensor
```
